### PR TITLE
[FEAT] ConformalSeasonalPool model (CSP)

### PR DESCRIPTION
**Pull request by nasaul**
## Summary

Implements the Conformal Seasonal Pool algorithm ([arxiv 2605.03789](https://arxiv.org/pdf/2605.03789)) as a new `ConformalSeasonalPool` model. CSP is a distribution-free, seasonality-aware probabilistic forecasting method that generates prediction intervals by mixing draws from two empirical pools — no parametric assumptions required.

This adds a third approach to uncertainty quantification in statsforecast, alongside the existing parametric (sigma-based) and conformal cross-validation methods:

| Approach | Examples | Mechanism |
|---|---|---|
| Parametric | `SeasonalNaive`, `ETS` | Normal quantiles from residual sigma |
| Conformal CV | `ConformalIntervals` | Rolling windows + absolute error quantiles |
| **CSP (new)** | `ConformalSeasonalPool` | Direct empirical sampling from two data pools |

## How it works

For each forecast horizon step, CSP draws `n_samples` from a mixture of:
- **Residual pool** (prob `1-w`): signed residuals `y[t] - y[t-m]` from the most recent `floor(calib_frac * n)` observations, added to the seasonal naive point forecast
- **Seasonal pool** (prob `w`): historical observations at the same seasonal phase, weighted by `exp(-λ * age)` where `age = (n-1) - t` so newer same-phase observations receive higher probability

The mixture weight `w` is computed **per horizon step** in the `"adaptive"` variant, based on the number of same-phase observations `k` available for that step:

```
m <= 1 (no seasonality)  → w = 0.0  (pure residual sampler)
k < 3   (thin history)   → w = 0.3
otherwise                → w = 0.5
```

Edge cases: `w` is clamped to `0.0` if `k == 0` (no pool obs for this phase) and to `1.0` if `R` is empty (no calibration residuals). The `"fixed"` variant always uses `w = 0.5`.

## Usage

```python
from statsforecast.models import ConformalSeasonalPool

# Adaptive variant (default) — fit/predict pattern
model = ConformalSeasonalPool(season_length=12)
model.fit(y).predict(h=12, level=[80, 95])
# → {"mean": ..., "lo-80": ..., "hi-80": ..., "lo-95": ..., "hi-95": ...}

# Stateless
ConformalSeasonalPool(season_length=12).forecast(y, h=12, level=[80, 95])

# Fixed variant
ConformalSeasonalPool(season_length=12, variant="fixed").forecast(y, h=12, level=[95])

# Raw samples with reproducible seed
model.simulate(h=12, n_paths=500, y=y, seed=42)  # → ndarray (500, 12)

# Works with StatsForecast
from statsforecast import StatsForecast
sf = StatsForecast(models=[ConformalSeasonalPool(season_length=12)], freq="ME")
sf.forecast(df=df, h=12, level=[80, 95])
```

## Parameters

| Parameter | Default | Description |
|---|---|---|
| `season_length` | required | Seasonal period `m` |
| `n_samples` | `100` | Samples used to estimate prediction intervals |
| `variant` | `"adaptive"` | `"adaptive"` or `"fixed"` mixture weight |
| `calib_frac` | `0.5` | Fraction of history used as calibration window (most recent `floor(calib_frac * n)` observations) |
| `decay` | `0.01` | Exponential recency rate λ for seasonal pool weights (`exp(-λ * age)`) |
| `alias` | auto | Defaults to `"CSP-Adaptive"` or `"CSP-Fixed"` |

## Implementation notes

- Point forecast (`"mean"`) is always the deterministic seasonal naive — not the sample mean — matching the paper's μ_h definition; uses `_seasonal_naive` internally
- When `n < season_length` (missing phases), NaN slots in μ are filled with `y[-1]` rather than propagated into the output
- `simulate()` accepts `error_distribution`/`error_params` for API compatibility but ignores them (emits a `UserWarning` if non-default values are passed)
- `predict_in_sample()` returns constant-width intervals derived from empirical quantiles of the calibration residuals R; returns NaN offsets when R is empty (`n == season_length`)
- No new dependencies — pure numpy
- Full `_TS` interface: `fit`, `predict`, `predict_in_sample`, `forecast`, `forward`, `simulate`

## Test plan

- [x] Smoke test: correct keys and shapes for all output dicts
- [x] Point forecast matches `SeasonalNaive` exactly
- [x] Interval ordering: `lo-95 ≤ lo-80` and `hi-80 ≤ hi-95`
- [x] `season_length=1` → `w=0`, pure residual sampler
- [x] Short series (< 3 cycles) → `w=0.3` in adaptive variant
- [x] CSP-Fixed and CSP-Adaptive produce different widths on short series
- [x] `simulate(seed=42)` is reproducible
- [x] `simulate(..., error_distribution="laplace")` emits `UserWarning`
- [x] `predict_in_sample` intervals are constant width
- [x] `fit/predict` point forecast matches `forecast()`
- [x] `h > season_length` repeats correctly
- [x] `n < season_length` (n ≤ m/2): no crash, all mean slots non-NaN after latest-obs fallback
- [x] `m/2 < n < m`: residual slice does not crash (negative index regression)
- [x] Empty R pool with non-empty per-phase pool produces non-zero interval width (regression for `k==0 and R.size==0` guard)
- [x] `predict_in_sample` with empty R returns NaN intervals (n == m edge case)

### CODE REVIEW

File: python/statsforecast/models.py [hunk 1]
```diff
@@ -4084,6 +4085,384 @@ def forward(
...
+        self.model_ = {
+            "y": y,
+            "calib_residuals": R,
+            "y_by_phase": y_by_phase,
+            "n_full_cycles": n // m,
+            "n": n,
+            "mean": mod["mean"],
+            "fitted": mod["fitted"],
+        }
+        return self
```
[the remark this answers is not in the document]
Comment (MMenchero) [reply python/statsforecast/models.py:4261]: Agree, we only need to store `{"y", "calib_residuals", "mean", "fitted"}`.

File: python/statsforecast/models.py [hunk 2]
```diff
@@ -4084,6 +4085,384 @@ def forward(
...
+            R = self.model_["calib_residuals"]
+            for lv in level:
+                lo_q = (1 - lv / 100) / 2
+                hi_q = 1 - lo_q
+                if R.size > 0:
+                    lo_offset = float(np.quantile(R, lo_q))
+                    hi_offset = float(np.quantile(R, hi_q))
+                else:
+                    lo_offset = hi_offset = np.nan
+                res[f"lo-{lv}"] = fitted + lo_offset
```
Comment (MMenchero) [inline python/statsforecast/models.py]: I think the standard convention, that comes from `add_fitted_pi`, is `fitted-lo-{lv}` and `fitted-hi-{lv}`.

File: tests/test_csp.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    assert any("CSP" in str(warning.message) for warning in w)
+
+
+# 4.10 — predict_in_sample structure and constant-width intervals
+def test_predict_in_sample(y_monthly):
+    model = ConformalSeasonalPool(season_length=12).fit(y_monthly)
+    res = model.predict_in_sample(level=[90])
+    assert "fitted" in res
+    assert "lo-90" in res
+    assert "hi-90" in res
```
Comment (MMenchero) [inline tests/test_csp.py]: If we change the name to `fitted-lo-{lv}}` and `fitted-hi-{lv}` to follow convention, then we also need to change the corresponding tests.

File: python/statsforecast/models.py [hunk 4]
```diff
@@ -4084,6 +4085,384 @@ def forward(
...
+            rng=rng,
+            mu=mu,
+        )
+        # Point forecast is seasonal naive. Using sample median instead would let the
+        # mixture draws shift the point estimate, but that diverges from the paper original goal: probabilistic forecasting.
+        res = {"mean": mu}
+        if level is not None:
+            level = sorted(level)
+            res.update(self._intervals_from_samples(samples, level))
+        return res
```
[the remark this answers is not in the document]
Comment (MMenchero) [reply python/statsforecast/models.py:4301]: CSP is a probabilistic model, so setting `level` to `None` will be unusual. Plus it's just slightly more expensive. No harm done.

File: python/statsforecast/models.py [hunk 5]
```diff
@@ -4084,6 +4085,384 @@ def forward(
...
+            rng=rng,
+            mu=mu,
+        )
+        res = {"mean": mu}
+        if level is not None:
+            level = sorted(level)
+            res.update(self._intervals_from_samples(samples, level))
+        if fitted:
+            res["fitted"] = mod["fitted"]
+        return res
```
[the remark this answers is not in the document]
Comment (MMenchero) [reply python/statsforecast/models.py:4385]: Same as before.

**Review by MMenchero — COMMENTED** [review]
Great job @nasaul I just left one small comment regarding naming convention and I think two of Copilot's suggestions are worth checking out.

File: python/statsforecast/models.py [hunk 6]
```diff
@@ -4084,6 +4085,379 @@ def forward(
...
+            alias (str, optional): Custom name of the model. Defaults to ``"CSP-Adaptive"`` or
+                ``"CSP-Fixed"`` based on ``variant``.
+        """
+        if variant not in ("adaptive", "fixed"):
+            raise ValueError("variant must be 'adaptive' or 'fixed'")
+        self.season_length = season_length
+        self.n_samples = n_samples
+        self.variant = variant
+        self.calib_frac = calib_frac
+        self.decay = decay
```
[the remark this answers is not in the document]
Comment (MMenchero) [reply python/statsforecast/models.py:4209]: The suggestion about `n_samples` could be useful, but the one about `season_length` should be done at a higher level for all models, if it not already implemented.

File: python/statsforecast/models.py [hunk 7]
```diff
@@ -4084,6 +4085,379 @@ def forward(
...
+        if k == 0:
+            w = 0.0
+        elif R.size == 0:
+            w = 1.0
+
+        if k > 0:
+            ages = (n - 1) - pool_idx
+            raw_w = np.exp(-decay * ages)
+            pool_weights = raw_w / raw_w.sum()
+            pool_draws = rng.choice(pool_vals, size=n_samples, p=pool_weights)
```
[the remark this answers is not in the document]
Comment (MMenchero) [reply python/statsforecast/models.py:4147]: This makes sense, but the first implementation of this model should stay as close as possible to the original source. So I suggest we skip this for now and wait for the author of the CSP model to make the final call.

**Review by MMenchero — DISMISSED** [review]
LGTM @nasaul. I added some comments to Copilot's suggestions, and other than that, I think we're ready for the author of the CSP model to review this first implementation.

**Comment by nasaul** [comment]
**Fixed — `ConformalSeasonalPool` prediction intervals were anti-conservative**

`_intervals_from_samples` and `predict_in_sample` both used plain `np.quantile` to compute `lo-L` / `hi-L` bounds from the predictive sample matrix and calibration residuals respectively. For a finite sample of `n` draws, plain quantile interpolation at level `q` is anti-conservative on the lower tail: it uses a fractional rank instead of the conformal-correct integer rank.

The fix applies the orientation correction described in Romano et al. — `floor((n+1)·q)/n` for lower-tail levels and `ceil((n+1)·q)/n` for upper-tail levels — via a new `_oriented_index` static method, applied in both interval-computing code paths.

A 197-series Monash benchmark measured a **1.7–1.8 pp coverage gap** at α=0.05 and α=0.10 before the fix (every gap significant at p < 1e-12). After the fix both paths agree to floating-point precision with the reference conformal implementation.

**Review by MMenchero — APPROVED** [review]
LGTM.
